### ISSUE TITLE
Transferred resolution does not equal device resolution

### ISSUE THREAD

**Issue by stoefln** [NetrisTV/ws-scrcpy#205]
I am trying to transfer the full device resolution (in a fork of your project), so I configured VideoSettings like this:

```
const currentSettings = new VideoSettings({
      lockedVideoOrientation: -1,
      bitrate: 524288,
      maxFps: 24,
      iFrameInterval: 1,
      bounds: new Size(3480, 3480),
      sendFrameMeta: false
    })
    
streamReceiver.sendMessage(CommandControlMessage.createSetVideoSettingsCommand(currentSettings))
```
I am expecting a resolution of **1080 x 2220** (Pixel 3A) and this is also what `streamReceiver.on('displayInfo', this.onDisplayInfo)` reports. 
BUT the decoded frames have a resolution of  **1072 x 2208**. 
### 🤷‍♂️
It would be great to know why scrcpy decides to send a lower resolution. Do you know?

**Comment by drauggres** [comment]
Hi.
Because of this two lines: https://github.com/NetrisTV/scrcpy/blob/451ebee56f9908edab8fa9d84232b0afb09da361/server/src/main/java/com/genymobile/scrcpy/ScreenInfo.java#L144-L145

AFAIK, h264 uses 8x8 pixels block size, this is why original scrcpy [has in this place](https://github.com/Genymobile/scrcpy/blob/ed84e18b1ae3e51d368f8c7bc88ba4db088e6855/server/src/main/java/com/genymobile/scrcpy/ScreenInfo.java#L111-L117) `~7` and resolition with it will be `1080x2216`

Broadway.js was the first decoded I integrated into this project and it works correctly only when dimensions are divided by 16. So I had to change the code to `~15` and this value stayed there since the first ws-scrcpy version. It is possible that other decoders would work with `~7`.

In theory, the correct way of handling this situation would be to use higher resolution (`1088x2224` in this case) and to tell decoder/renderer to skip unused pixels, but I haven't dig deeper.

**Comment by stoefln** [comment]
Thanks a lot, @drauggres, that helps a lot already. I saw that scrcpy also provides a `crop` option (oh, and VideoSettings also offers a crop property) which would be ideal to get rid of the empty pixels, right?
So I just need to somehow tell scrcpy to output 1088x2224, but to not scale up the original resolution.

**Comment by drauggres** [comment]
I never used `crop`, but I believe it is used to cut from original image, i.e. before encoding.

I tried to remove rounding completely and was able to get `1080x2220` video with h264-converter (but it broke broadway.js). So I guess it all comes down to decoder.

If you want to debug scrcpy-server.jar [here](https://github.com/NetrisTV/ws-scrcpy/blob/master/docs/debug.md#android-server-scrcpy-serverjar) is the instruction (don't forget to change version in the command).
For original scrcpy: [DEVELOP.md](https://github.com/Genymobile/scrcpy/blob/master/DEVELOP.md#debug-the-server)

**Comment by stoefln** [comment]
Great, thanks, will try tomorrow to transfer 1088x2224 and crop it on the client

**Comment by stoefln** [comment]
@drauggres I think I found a setting for making this work: I would like to try to set
`codec.setVideoScalingMode(MediaCodec.VIDEO_SCALING_MODE_SCALE_TO_FIT_WITH_CROPPING);` [here](https://github.com/NetrisTV/scrcpy/blob/451ebee56f9908edab8fa9d84232b0afb09da361/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java#L111)
(Default is `MediaCodec.VIDEO_SCALING_MODE_SCALE_TO_FIT`)

I followed the build steps [here](https://github.com/Genymobile/scrcpy/blob/master/BUILD.md#mac-os), but unfortunately I am getting a lot of errors like

```
 scrcpy git:(feature/websocket-server) ✗ ninja -Cx
ninja: Entering directory `x'
[1/55] Compiling C object app/scrcpy.p/src_compat.c.o
FAILED: app/scrcpy.p/src_compat.c.o
cc -Iapp/scrcpy.p -Iapp -I../app -I../app/src -I/opt/homebrew/Cellar/ffmpeg@4/4.4.2/include -I/opt/homebrew/include/SDL2 -flto -fcolor-diagnostics -Wall -Winvalid-pch -Wextra -std=c11 -O0 -g -D_THREAD_SAFE -MD -MQ app/scrcpy.p/src_compat.c.o -MF app/scrcpy.p/src_compat.c.o.d -o app/scrcpy.p/src_compat.c.o -c ../app/src/compat.c
In file included from ../app/src/compat.c:1:
../app/src/compat.h:12:10: fatal error: 'SDL2/SDL_version.h' file not found
#include <SDL2/SDL_version.h>
         ^~~~~~~~~~~~~~~~~~~~
```
Any idea?

**Comment by stoefln** [comment]
Nevermind! Just found a way to compile. I downloaded the sdl sources and copied the header files into the project.

**Comment by stoefln** [comment]
But now I got an apk file 🤔
![image](https://user-images.githubusercontent.com/562800/174570594-f68f60b2-e737-46d7-b1ab-fe72eb0d22b5.png)

**Comment by drauggres** [comment]
> I don't seem to manage to start the apk via `CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process -agentlib:jdwp=transport=dt_socket,suspend=y,server=y,address=5005 / com.genymobile.scrcpy.Server 1.19-ws2 DEBUG web 8886`

Are you running it on Android 8 or below? For 9+ use other arguments.
Did you forward `5005` port (`adb forward tcp:5005 tcp:5005`)?

>  The line seems to set a classpath, not quite sure: Is this indeed supposed to start the jar (which is actually an apk)?

Yes, it should start the server, but also it will pause execution until the debugger is connected (`suspend=y` in the command).

It is possible that something already using `5005`, you can change it to any other.

**Comment by stoefln** [comment]
@drauggres Oh, forgot to forward. But I sent the right command, the one for Android 9+.
Here is what I execute (now after changing debug port to 5006): 
`CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process -XjdwpProvider:internal -XjdwpOptions:transport=dt_socket,suspend=y,server=y,address=5006 / com.genymobile.scrcpy.Server 1.19-ws2 web DEBUG 8886`
Then I do the port forward:
`adb forward tcp:5006 tcp:5006`
But `adb shell pidof app_process`does not show anything.
And when I try to debug via AS, I get this:
<img width="953" alt="Screenshot 2022-06-20 at 16 31 43" src="https://user-images.githubusercontent.com/562800/174626464-6b05d065-1661-4978-9dcf-1d2b2a660159.png">

**Comment by drauggres** [comment]
Is there anything in `logcat`?

Maybe it is a device specific problem. And there was a problem with debugging on some of the latest Android versions (don't recall was it 12 or 13 beta). You should try to debug with emulator and Android 10-11.

P.S. you didn't forgot to do `adb push server/build/outputs/apk/debug/server-debug.apk /data/local/tmp/scrcpy-server.jar`, right?

**Comment by stoefln** [comment]
@drauggres thanks so much for your help. I didn't manage to solve the debugger issue, but I managed to compile and run the apk using JS.
I also managed to introduce black bars in order to preserve aspect ratio and resolution of the video. See changes here:
https://github.com/NetrisTV/scrcpy/compare/feature/websocket-server...repeato-qa:feature/keep-aspect-ratio

Let me know if I can return a favor!
https://github.com/repeato-qa/scrcpy/commit/f2d21b8a2d1827f4fdebf7cfd0a412ec7411beff
